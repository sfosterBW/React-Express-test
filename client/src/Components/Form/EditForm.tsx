import React, { FC, FormEvent } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { InputCheckbox, InputText } from '../Input/Input'
import useField from '../../utils/hooks'
import { Fruit } from '../../utils/interfaces'
import { RootState } from '../../Reducers/store'
import { toggleAlert } from '../../Reducers/alertReducer'
import { updateFruit } from '../../Reducers/fruitReducer'
import { toggleModal } from '../../Reducers/modalReducer'
import styles from './Form.module.css'

interface Props {
  fruit: Fruit
  title?: string
}

const EditForm: FC<Props> = ({ fruit, title = "Edit a fruit" }) => {
  const dispatch = useDispatch()

  const modal = useSelector((state: RootState) => state.modal.toggle)

  const [best, resetBest] = useField<boolean>("Is it best?", "best", "checkbox", fruit.best)
  const [description, resetDescription] = useField<string>("Description", "description", "text", "")
  const [name, resetName] = useField<string>("Add a fruit", "name", "text", fruit.name)
  
  const submit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
    event.preventDefault()
    dispatch(updateFruit({
      id: fruit.id,
      best: best.value,
      name: name.value,
      description: description.value === "" ? undefined : description.value
    }))
    dispatch(toggleAlert(`${name.value} has been updated`, true))

    if (modal) {
      dispatch(toggleModal(false))
    }

    resetBest()
    resetDescription()
    resetName()
  }

  return (
    <form className={styles.form} onSubmit={(event) => submit(event)}>
      <h2 className={styles.title}>
        {title}
      </h2>
      <InputText {...name} />
      <InputText {...description} />
      <InputCheckbox {...best} />
      <button className={styles.button} data-testid="submit-edit">
        Edit fruit
      </button>
    </form>
  )
}

export default EditForm
