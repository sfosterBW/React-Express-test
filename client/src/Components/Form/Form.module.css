.form {
  padding: 4vmin 6vmin;
  border-radius: 0.5vmin;
  background: #f0bbc5b5;
  margin: 3vmin;
}

.title {
  font-size: calc(16px + 2vmin);
  text-align: center;
}

.button {
  display: flex;
  margin: 0 auto;
  padding: 2vmin 4vmin;
  font-size: calc(12px + 1vmin);
  font-weight: bold;
  color: white;
  text-shadow: 0.2vmin 0.2vmin gray;
  border-radius: 0.5vmin;
  background: #1ab545cc;
}

.button:hover {
  position: relative;
  top: -0.1vmin;
  left: -0.1vmin;
  box-shadow: 0 0.5vmin 0.5vmin rgba(0, 0, 0, 0.75);
}

.button:active {
  position: relative;
  top: 0.1vmin;
  left: 0.1vmin;
}
