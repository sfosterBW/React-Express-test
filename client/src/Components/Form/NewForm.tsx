import React, { FC, FormEvent } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { InputCheckbox, InputText } from '../Input/Input'
import useField from '../../utils/hooks'
import { RootState } from '../../Reducers/store'
import { toggleAlert } from '../../Reducers/alertReducer'
import { createFruit } from '../../Reducers/fruitReducer'
import { toggleModal } from '../../Reducers/modalReducer'
import styles from './Form.module.css'

const NewForm: FC<{ title?: string }> = ({ title = 'Add a new fruit' }) => {
  const dispatch = useDispatch()

  const modal = useSelector((state: RootState) => state.modal.toggle)

  const [best, resetBest] = useField<boolean>('Is it best?', 'best', 'checkbox', false)
  const [description, resetDescription] = useField<string>('Description', 'description', 'text', '')
  const [name, resetName] = useField<string>('Add a fruit', 'name', 'text', '')

  const submit = (event: FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    dispatch(createFruit({
      name: name.value,
      description: description.value === '' ? undefined : description.value,
      best: best.value
    }))
    dispatch(toggleAlert(`${name.value} has been added`, true))

    if (modal) {
      dispatch(toggleModal(false))
    }

    resetName()
    resetDescription()
    resetBest()
  }

  return (
    <form className={styles.form} onSubmit={(event): void => submit(event)}>
      <h2 className={styles.title} data-testid="new-form-title">
        {title}
      </h2>
      <InputText {...name} />
      <InputText {...description} />
      <InputCheckbox {...best} />
      <button className={styles.button} data-testid="new-form-submit">
        Add new fruit
      </button>
    </form>
  )
}

export default NewForm
