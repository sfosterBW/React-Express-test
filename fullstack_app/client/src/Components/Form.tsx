import React, { FC, useState } from 'react'

interface Props {
  best?: boolean
  handleSubmit: (id: number, name: string, best: boolean) => void
  id?: number
  name?: string
}

const Form: FC<Props> = ({
  best = false,
  handleSubmit,
  id = -1,
  name = "" }) => {

  const [bestInput, setBest] = useState<boolean>(best)
  const [nameInput, setName] = useState<string>(name)

  const reset = () => {
    setBest(best)
    setName(name)
  }

  const setter = (set: any) =>
    (event: any) => {
      const { checked, value } = event.target
      const newValue = checked ? checked : value
      set(newValue)
    }

  const submit = (event: any) => {
    event.preventDefault()
    if (nameInput.length > 0) {
      handleSubmit(id, nameInput, bestInput)
      reset()
    } else {
      alert("You need to add a name")
    }
  }

  return (
    <form className="fruit-form" onSubmit={(event: any) => submit(event)}>
      <h2>List out your favourite fruit</h2>
      <div className="input-wrapper">
        <label htmlFor="name">Add a fruit:</label>
        <input
          name="name"
          onChange={setter(setName)}
          type="text"
          value={nameInput} />
      </div>
      <div className="input-wrapper">
        <label htmlFor="best">Best:</label>
        <input
          checked={bestInput}
          name="best"
          onChange={setter(setBest)}
          type="checkbox"
          value={bestInput.toString()} />
      </div>
      <button>Add new fruit</button>
    </form>
  )
}

export default Form
