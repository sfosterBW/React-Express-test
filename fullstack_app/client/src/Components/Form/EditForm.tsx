import React, { FC, FormEvent, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { InputCheckbox, InputText } from '../Input/Input'
import { useField } from '../../utils/hooks'
import { Fruit } from '../../utils/interfaces'
import { RootState } from '../../Reducers/store'
import { toggleAlert } from '../../Reducers/alertReducer'
import { updateFruit } from '../../Reducers/fruitReducer'
import { toggleModal } from '../../Reducers/modalReducer'
import styles from './Form.module.css'

interface Props {
  fruit: Fruit
  title?: string
}

const EditForm: FC<Props> = ({ fruit, title = "Edit a fruit" }) => {
  const [best, setBest] = useState<boolean>(fruit.best)
  const name = useField("Add a fruit", "name", "text", fruit.name)
  const toggle = (state: RootState) => state.modal.toggle
  const modal = useSelector(toggle)
  const dispatch = useDispatch()

  const setter = (set: any) => (
    (event: any): void => {
      const { checked, type, value } = event.target
      const newValue = type === "checkbox" ? checked : value
      set(newValue)
    }
  )

  const submit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
    event.preventDefault()

    dispatch(updateFruit({
      id: fruit.id,
      best: best,
      name: name.value,
    }))
    dispatch(toggleAlert(`${name.value} has been updated`, true))

    if (modal) {
      dispatch(toggleModal(false))
    }
    
    name.reset()
    setBest(false)
  }

  return (
    <form className={styles.form} onSubmit={(event) => submit(event)}>
      <h2 className={styles.title}>{title}</h2>
      <InputText {...name} />
      <InputCheckbox
        checked={best}
        handleChange={setter(setBest)}
        label="Is it best?"
        name="best" />
      <button className={styles.button}>Edit fruit</button>
    </form>
  )
}

export default EditForm
