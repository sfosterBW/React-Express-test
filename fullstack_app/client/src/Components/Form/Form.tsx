import React, { FC, FormEvent, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import fruitService from '../../utils/api'
import { InputCheckbox, InputText } from '../Input/Input'
import { useField } from '../../utils/hooks'
import { IFruit } from '../../utils/interfaces'
import { RootState } from '../../utils/store'
import { createFruit, toggleAlert, toggleModal, updateFruit } from '../../utils/actions'
import styles from './Form.module.css'

interface Props {
  fruit?: IFruit
  title?: string
}

const defaultFruit = { id: undefined, name: "", best: false }

const Form: FC<Props> = ({ fruit = defaultFruit, title = "Add a new fruit" }) => {
  const [best, setBest] = useState<boolean>(fruit.best)
  const name = useField(fruit.name, "Add a fruit", "name", "text")
  const selectModal = (state: RootState) => state.modal.toggle
  const modal = useSelector(selectModal)
  const dispatch = useDispatch()

  const setter = (set: any) =>
    (event: any) => {
      const { checked, type, value } = event.target
      const newValue = type === "checkbox" ? checked : value
      set(newValue)
    }

  const handleSubmit = async (fruit: IFruit) => {
    if (fruit.id) {
      const res = await fruitService.updateFruit(fruit)
      if(res.status === 201) {
        dispatch(updateFruit(res.data))
      } else {
        dispatch(toggleAlert(true))
      }
    } else {
      const res = await fruitService.createFruit(fruit)
      if(res.status === 201) {
        dispatch(createFruit(res.data))
      } else {
        dispatch(toggleAlert(true))
      }
    }
    modal && dispatch(toggleModal(false))
  }

  const submit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    if (name.value.length > 0) {
      handleSubmit({
        id: fruit.id,
        best: best,
        name: name.value,
      })
      name.reset()
      setBest(false)
    } else {
      dispatch(toggleAlert(true))
    }
  }

  return (
    <form className={styles.form} onSubmit={(event) => submit(event)}>
      <h2 className={styles.title}>{title}</h2>
      <InputText {...name} />
      <InputCheckbox
        checked={best}
        handleChange={setter(setBest)}
        label="Is it best?"
        name="best" />
      <button className={styles.button}>Add new fruit</button>
    </form>
  )
}

export default Form
