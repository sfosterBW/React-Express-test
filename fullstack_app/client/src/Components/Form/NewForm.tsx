import React, { FC, FormEvent, useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import fruitService from '../../utils/api'
import { InputCheckbox, InputText } from '../Input/Input'
import { useField } from '../../utils/hooks'
import { NewFruit } from '../../utils/interfaces'
import { RootState } from '../../utils/store'
import { createFruit, toggleAlert, toggleModal } from '../../utils/actions'
import styles from './Form.module.css'

const NewForm: FC<{title?: string}> = ({ title = "Add a new fruit" }) => {
  const [best, setBest] = useState<boolean>(false)
  const name = useField("", "Add a fruit", "name", "text")
  const toggle = (state: RootState) => state.modal.toggle
  const modal = useSelector(toggle)
  const dispatch = useDispatch()

  const setter = (set: any) => (
    (event: any) => {
      const { checked, type, value } = event.target
      const newValue = type === "checkbox" ? checked : value
      set(newValue)
    }
  )

  const handleSubmit = async (fruit: NewFruit): Promise<void> => {
    try{
      const newFruit = await fruitService.createFruit(fruit)
      dispatch(createFruit(newFruit))
    }
    catch (error) {
      console.log(error)
      dispatch(toggleAlert('submit error', true))
    }
    modal && dispatch(toggleModal(false))
  }

  const submit = (event: FormEvent<HTMLFormElement>): void => {
    event.preventDefault()
    if (name.value.length > 0) {
      const newFruit: NewFruit = { name: name.value, best }
      handleSubmit(newFruit)
      name.reset()
      setBest(false)
    } else {
      dispatch(toggleAlert('submit error', true))
    }
  }

  return (
    <form className={styles.form} onSubmit={(event) => submit(event)}>
      <h2 className={styles.title}>{title}</h2>
      <InputText {...name} />
      <InputCheckbox
        checked={best}
        handleChange={setter(setBest)}
        label="Is it best?"
        name="best" />
      <button className={styles.button}>Add new fruit</button>
    </form>
  )
}

export default NewForm
