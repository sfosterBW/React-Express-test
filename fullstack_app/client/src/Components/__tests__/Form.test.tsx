import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'

import Form from '../Form'

describe('the Form component', () => {

  describe('without props', () => {

    const mockFunction = jest.fn()
    const formComponent = <Form handleSubmit={mockFunction()} />
    const form = mount(formComponent)

    it('renders with the correct structure', () => {
      expect(form).toBeDefined()
      expect(form.find('form')).toHaveLength(1)
      expect(form.find('h2')).toHaveLength(1)
      expect(form.find('InputText')).toHaveLength(1)
      expect(form.find('InputCheckbox')).toHaveLength(1)
      expect(form.find('button')).toHaveLength(1)
    })

    it('renders the same as last time', () => {
      const tree = renderer
        .create(formComponent)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

  })

  describe('with props', () => {

    const testFruit = { _id: 2, name: "Banana", best: true }
    const mockFunction = jest.fn()
    const formComponent = <Form
      fruit={testFruit}
      handleSubmit={mockFunction()} />
    const form = mount(formComponent)

    it('functions properly with props', () => {
      expect(form).toBeDefined()
      expect(form.find('InputText').props().value).toEqual(testFruit.name)
      expect(form.find('InputCheckbox').props().checked).toEqual(testFruit.best)
      form.find('button').simulate('click')
      expect(mockFunction).toHaveBeenCalled()
    })

    it('renders the same as last time', () => {
      const tree = renderer
        .create(formComponent)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
