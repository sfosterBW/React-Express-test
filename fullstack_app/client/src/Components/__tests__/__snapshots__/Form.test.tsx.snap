// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Form component with custom props renders the same as last time 1`] = `
<form
  className="fruit-form"
  onSubmit={[Function]}
>
  <h2>
    List out your favourite fruit
  </h2>
  <div
    className="input-wrapper"
  >
    <label
      htmlFor="name"
    >
      Add a fruit:
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value="Banana"
    />
  </div>
  <div
    className="input-wrapper"
  >
    <label
      htmlFor="best"
    >
      Is it best?:
    </label>
    <input
      checked={true}
      name="best"
      onChange={[Function]}
      type="checkbox"
      value="true"
    />
  </div>
  <button>
    Add new fruit
  </button>
</form>
`;

exports[`the Form component with default props renders the same as last time 1`] = `
<form
  className="fruit-form"
  onSubmit={[Function]}
>
  <h2>
    List out your favourite fruit
  </h2>
  <div
    className="input-wrapper"
  >
    <label
      htmlFor="name"
    >
      Add a fruit:
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="input-wrapper"
  >
    <label
      htmlFor="best"
    >
      Is it best?:
    </label>
    <input
      checked={false}
      name="best"
      onChange={[Function]}
      type="checkbox"
      value="false"
    />
  </div>
  <button>
    Add new fruit
  </button>
</form>
`;
